src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/CourtCaseEventsService.kt

```diff
@@ -28,78 +33,55 @@ class CourtCaseEventsService(
   fun processPersonFromCourtCaseEvent(person: Person) {
     log.debug("Entered processPersonFromCourtCaseEvent")
     if (person.otherIdentifiers?.pncIdentifier?.pncId.isNullOrEmpty()) {
-      telemetryService.trackEvent(TelemetryEventType.MISSING_PNC, emptyMap())
+      trackEvent(MISSING_PNC, emptyMap())
       return
     }
     person.otherIdentifiers?.pncIdentifier?.let { pncIdentifier ->
       val pncId = pncIdentifier.pncId
       if (pncIdentifier.isValid()) {
-        telemetryService.trackEvent(TelemetryEventType.VALID_PNC, mapOf("PNC" to pncIdentifier.toString()))
+        trackEvent(VALID_PNC, mapOf("PNC" to pncIdentifier.toString()))
         val defendants = personRepository.findByDefendantsPncNumber(PNCIdentifier(pncId))?.defendants.orEmpty()
-        if (matchesExistingRecordExactly(defendants, person)) {
-          log.info("Exactly matching Person record exists with defendant - no further processing will occur")
-          telemetryService.trackEvent(
-            TelemetryEventType.NEW_CASE_EXACT_MATCH,
-            mapOf("PNC" to pncId, "CRN" to defendants[0].crn, "UUID" to person.personId.toString()),
-          )
-        } else if (matchesExistingRecordPartially(defendants, person)) {
-          log.info("Partially matching Person record exists with defendant - no further processing will occur")
-          telemetryService.trackEvent(
-            TelemetryEventType.NEW_CASE_PARTIAL_MATCH,
-            extractMatchingFields(defendants[0], person),
-          )
-        } else if (defendants.isEmpty()) {
-          log.debug("No existing matching Person record exists - creating new person and defendant with pnc $pncId")
-          val personRecord = personRecordService.createNewPersonAndDefendant(person)
-          personRecord.let {
-            offenderService.processAssociatedOffenders(it, person)
-            prisonerService.processAssociatedPrisoners(it, person)
-          }
+        val defendantMatcher = DefendantMatcher(defendants, person)
 
-          telemetryService.trackEvent(
-            NEW_CASE_PERSON_CREATED,
-            mapOf("UUID" to personRecord.personId.toString(), "PNC" to pncId),
-          )
-        } // what if defendants is not empty?
+        when {
+          defendantMatcher.isExactMatch() -> exactMatchFound(defendantMatcher, person, pncId)
+          defendantMatcher.isPartialMatch() -> partialMatchFound(defendantMatcher)
+          else -> {
+            createNewPersonRecordAndProcess(person, pncId)
+          }
```

Comment:
Hope this works without causing any issues. 🤞

Reply:
I think it will be fine - the way the old code was written led you to believe that there was a possibility that non-matching defendants had been found by `personRepository.findByDefendantsPncNumber(PNCIdentifier(pncId))?.defendants.orEmpty()`. 
I don't think that is possible.